Endpoint redirect cleanup

jrajahalme opened the pull request:
Rationalize sequence of events for policy update and applying incremental policy map updates:

- update proxy network policy right after `ConsumeMapChanges` in `applyPolicyMapChanges` so that it is more obvious that the selector cache version for bpf map updates and proxy network policy is the same on (as updated by ConsumeMapChanges)
- defer applying the incremental changes for the full policy map sync when a new policy has been computed, so that the proxy has had time to configure the new listeners and to apply the new network policy before getting traffic for them.
- create l7 redirects before incremental updates, so that the proxy ports are available for them

Keep realized redirects in EndpointPolicy:
- can create redirects before holding endpoint lock, no coordination needed between redirects of different policy instances (old vs. new)
- Create MapState after redirects so that proxy ports are available during MapState generation removing the need for a second pass to patch up proxy ports
- Move old redirect removal to happen after proxy ACKs the changes, removing the need for the 1st maps sync, as now the main map sync happens before redirects are removed

jrajahalme committed:
endpoint: Update network policy after incremental updates

Check for the need to update network policy after applying incremental
updates. This makes sure network policy update has access to the selector
cache version used for the incremental desired policy updates.

If there is no need for network policy update, still wait for any ongoing
updates to have finished as the current state may depend on them
finished.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
endpoint: Get DNS restore rules after policy update

Now that redirects are added before taking the endpoint lock we can
update DNS restore rules with the new policy rules after the policy
update, rather than from the old rules before the policy rules updates.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -679,11 +679,68 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext, rul
+		err = e.applyPolicyMapChanges(regenContext, e.desiredPolicy != e.realizedPolicy)
+		if err != nil {
+			return err
+		}
```
I see that the code changed is reordering the logic around, how do we know that making this change is safe in this order? Is this a bug fix? I'm not sure I fully understand the motivation for this change. 

Also, have you looked at the locking patterns now that the order has changed? Often times we introduce deadlocks into this code because we move things around and don't realize that there was some sort of lock ordering dependency.

jrajahalme replied:
1st commit only refactors operations happening within the same endpoint lock, so there is no locking impact. It moves calls to `e.updateAndOverrideEndpointOptions(nil)` up so that it can be called from one code path rather than many. `addNewRedirects` is moved to happen before `applyPolicyMapChanges` so that the proxy ports exist when creating redirect map state entries. Some logging is moved to `applyPolicyMapChanges`, still under the same conditions. `updateNetworkPolicy` is moved into `applyPolicyMapChanges`, to the end, so that it still happens after the desiredPolicy has been updated by `ConsumeMapChanges` before the proxy network policy is updated.

There is one possibly observable change in the 1st commit: The bpf map updates via `applyPolicyMapChanges` are skipped, and are instead done by the following full map sync when a new policy has been computed. Previously is was possible that e.g., new redirects were pushed to the bpf maps by `applyPolicyMapChanges` when Envoy was still processing the new listeners and/or the network policy for them. This could have resulted in "policy not found" type of errors in the proxy, if traffic was redirected to the (new) listeners for endpoints whose policy was not yet updated.

The main change is in the 2nd commit, where `addNewRedirects` is computed earlier, between the selector policy update and creation of the new map state. Just checked for locking requirements, and I was missing the endpoint lock there! It looks like read lock suffices, though. This code path calls in to `pkg/proxy`, which may call back to `endpoint` via `EndpointUpdater`.

christarazi replied:
That sounds reasonable to me; this would be great to include in the commit msg itself :)

christarazi reviewed (commented):
The overall change LGTM, I think it simplifies things. I have one comment below as these things tend to be very subtle and just want to ensure we get it right. Typically, I'd expect `cleanup` label to indicate changes without a functional impact, but this PR does have functional impact, right?

jrajahalme committed:
endpoint: Skip incremental policy map updates for new policy

When full policy update path has computed a new endpoint policy, skip
applying incremental policy map updates after applying the incremental
changes to endoint's desired policy, as a full desired/realized map sync
is to be done after Envoy has acknowledged new listeners and network
policy.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1129,60 +1097,104 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
+		return nil
+	}
+
+	if hasNewPolicy {
```
bad merge? this conditional is duplicated.

jrajahalme replied:
yes, thanks for noticing! FIxed.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1129,60 +1097,104 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
+			Debug("applyPolicyMapChanges: Updating Envoy NetworkPolicy")
+		stats.proxyPolicyCalculation.Start()
+		var rf revert.RevertFunc
+		err, rf = e.updateNetworkPolicy(proxyWaitGroup)
```
Idea: rename `updateNetworkPolicy()` to `updateProxyPolicy()`? And maybe we can just get rid of `updateNetworkPolicy()` entirely?

squeed replied:
Further commentary: the DNS proxy doesn't care about incremental updates, so the assumption that "proxy is envoy" is true here. However, the method names do not make this distinction, and the code generally multiplexes Envoy vs. DNS in `pkg/proxy/Proxy` rather than `pkg/endpoint/Endpoint`

Does it make sense to push any of the "should we update Envoy?" predicates in to `pkg/proxy`? It all seems a bit awkward for endpoints to care about Envoy in particular.

jrajahalme replied:
Got rid of it! Better not give chances calling it anywhere else...

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1129,60 +1097,104 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
+		e.getLogger().
+			WithField(logfields.SelectorCacheVersion, e.desiredPolicy.VersionHandle).
+			Debug("applyPolicyMapChanges: Using current Networkpolicy")
+		e.useCurrentNetworkPolicy(proxyWaitGroup)
```
AWESOME! :-) It's really nice to have this race fixed up.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
-			// the caller.
-			if e.desiredPolicy == e.realizedPolicy {
-				return nil
+			pp := e.newProxyPolicy(l4, v, dstPort, dstProto)
```
`e.newProxyPolicy()` does not reference `e` at all. Can we get rid of it? In fact, can we get rid of that type altogether, and move it to `pkg/proxy`?

jrajahalme replied:
This type replaced passing `*policy.L4Filter` directly to the proxy package, and was implemented at the endpoint package due to named port resolution happening there. Having the endpoint as a receiver for `newProxyPolicy` is likely an historical artifact, removed. IMO moving the implementation to the proxy package would be odd as it uses this via `endpoint.EndpointUpdater`, expecting the implementation to be outside of the package.

squeed replied:
Reasonable enough. Thanks for the cleanup.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
-				revertStack.Push(revertFunc)
-				desiredRedirects[proxyID] = redirectPort
+	selectorPolicy.CreateRedirects(func(l4 *policy.L4Filter) error {
+		for _, v := range l4.PerSelectorPolicies {
```
nit: this function is long enough that `v` is not a meaningful name. can you pick a better one?

jrajahalme replied:
renamed as `perSelectorPolicy`

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
-				finalizeList.Append(finalizeFunc)
-				revertStack.Push(revertFunc)
-				desiredRedirects[proxyID] = redirectPort
+	selectorPolicy.CreateRedirects(func(l4 *policy.L4Filter) error {
```
`pkg/policy/SelectorPolicy.CreateRedirects()` seems to be ripe for refactor. It.. doesn't create redirects, it just loops over all L4Filters. The only caller takes the supplied L4Filter and iterates over all PerSelectorPolicies.

Can we change this to a `SelectorPolicy.L4Filters() iter.Seq[*policy.L4Filter]`?

jrajahalme replied:
I guess I intended to move the ownership of the resulting slice of redirects to the policy package, but that did not materialize. Let me wee how I can simplify this...

jrajahalme replied:
a bit simpler with `SelectorPolicy.RedirectFilters() iter.Seq2[*L4Filter, *PerSelectorPolicy]`

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
+			// desiredRedirects starts out empty, so we can use it check
+			// if the redirect has already been updated on this round.
+			if desiredRedirects[proxyID] != 0 {
+				continue
```
if `e.proxy.CreateOrUpdateRedirect()` have access to both the `L4Policy` and the `PerSelectorPolicy`, why is it safe to skip here? This seems like it could cause breakage -- what if `CreateOrUpdateRedirect()` reads from more fields than is in the `proxyID`? We could have a misconfiguration.

squeed replied:
I suspect we need to refactor `CreateOrUpdateRedirect()` to get everything it needs from the ProxyID?

jrajahalme replied:
`PerSelectorPolicy` is only used to access the listener name, which is also part of the `proxyID`. I'll change the `proxyPolicy` implementation to make this clear.

squeed replied:
much cleaner, thanks!

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -318,78 +307,27 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
-		return nil, nil
+// Must be called with endpoint.mutex locked for writing, as this calls back to
+// 'e.OnDNSPolicyUpdateLocked()'.
+func (e *Endpoint) removeOldRedirects(desiredRedirects, realizedRedirects map[string]uint16) {
```
nit: maybe name the arguments `newRedirects, oldRedirects`?

jrajahalme replied:
That would read like the `oldRedirects` would need to be removed? Id prefer to let this be for now.

squeed replied:
Sure, just a thought.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
-				// redirect for that port.
-				statsKey := policy.ProxyStatsKey(l4.Ingress, string(l4.Protocol), dstPort, redirectPort)
-				proxyStats := e.getProxyStatistics(statsKey, string(l4.L7Parser), dstPort, l4.Ingress, redirectPort)
+			var redirectPort uint16
```
nit: this variable seems redundant.

jrajahalme replied:
removed, thanks!

squeed commented on pkg/endpoint/policy.go:
```diff
@@ -258,15 +249,46 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics) (*policyGener
+	// Ingress endpoint needs no redirects
+	if !e.isProperty(PropertySkipBPFPolicy) {
+		stats.proxyConfiguration.Start()
+		desiredRedirects, ff, rf = e.addNewRedirects(result.selectorPolicy, datapathRegenCtxt.proxyWaitGroup)
```
The vague distinction between `e.regeneratePolicy()` and `e.setDesiredPolicy()` was that `regeneratePolicy()` shouldn't mutate the endpoint, since it drops the endpoint lock for all of the heavy lifting.

I see the issue, though -- we need to know the `proxyPort` to generate the MapState.

Let me think if there's a reasonably clever way to work around this. If not, then c'est la vie.

jrajahalme replied:
There is a bigger refactor change endpoints to share redirects for the same proxy type and port (underlying proxy listener is already shared), this would move redirect management to policy ingestion and we could fail out on policy ingestion if proxy redirect for it can not be created. One significant obstacle right now is that `pkg/proxy` imports `pkg/policy` and hence policy package can not use the types or functions of the proxy package, at least not directly.

Then, from the viewpoint of the endpoint, if the policy needs redirects, they would already exist and the policy package could place in the right proxy ports without the endpoint package being involved.

However, this is out of the scope for this PR, IMO.

jrajahalme replied:
This is seemingly blocked by the named port being per-endpoint, but that can be likely separated from the *creation* of the redirect, if per-endpoint policy applicable to the redirects is configured separately.

squeed replied:
The reality is that calculating policy is highly side-effecting anyways, so the distinction is extremely academic anyways. As long as we hold the endpoint lock for as short a time as possible, I’m satisfied.

jrajahalme replied:
So we're all good for this PR then?

squeed replied:
Yup :-)

squeed commented on pkg/endpoint/policy.go:
```diff
@@ -278,7 +300,20 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics) (*policyGener
 // relying on.
-func (e *Endpoint) setDesiredPolicy(res *policyGenerateResult) error {
+func (e *Endpoint) setDesiredPolicy(res *policyGenerateResult, datapathRegenCtxt *datapathRegenerationContext) error {
+	defer func() {
```
this will panic if `res == nil`.

Suggestion: just move this to the `if e.identityRevision != res.identityRevision {` block; it's the only case where this function exits early.

jrajahalme replied:
Done :-)

squeed commented on pkg/proxy/proxy.go:
```diff
@@ -830,72 +819,54 @@ func (p *Proxy) revertStackUnlocked(revertStack revert.RevertStack) {
 
 // RemoveRedirect removes an existing redirect that has been successfully created earlier.
-func (p *Proxy) RemoveRedirect(id string, wg *completion.WaitGroup) (error, revert.FinalizeFunc, revert.RevertFunc) {
+// Called with 'localEndpoint' passed to 'CreateOrUpdateRedirect' locked for writing!
```
We only write so we can clean up some DNS rules; maybe we can get rid of this codepath somehow. I'll think about it. (doesn't need to be part of this PR unless inspiration strikes you)

squeed replied:
Oh, I think we can move DNS rules caching to its own controller; we should just periodically synchronize them. I'll take this as a separate task.

jrajahalme committed:
endpoint: Add redirects before MapState generation

Move the set of desired redirects to the EndpointPolicy, as it is stable
for any computed EndpointPolicy. If some redirects are missing, e.g., due
to missing named ports or Envoy listeners, the policy will be recomputed
when those are added, and the new policy will then contain also those
redirects.

This also allows each EndpointPolicy instance to have it's own set of
redirects, depending on the selectorPolicy. If we kept a single set of
realized redirects for the endpoint, we'd need to combine the current and
new sets for the redirects to be available when incremental updates are
added both to the current and new EndpointPolicy.

Move adding redirects to happen before the MapState for the new
EndpointPolicy is computed. This way the redirects are available when
needed for MapState generation, removing the need for the second pass to
add the redirects to the MapState. Redirect creation is simplified with
new SelectorPolicy.RedirectFilters() iterator.

Eliminate the (first) bpf policy map sync that was done just before
removing old redirect.  Old redirects are now removed in a finalizer
function, that is run after we get acknowledgements for the newly added
redirects (if any). If there is a NACK instead, then the finalizer
function is not called at all, so there is no need for reverting the
removals.

The first sync could have sent traffic to Envoy while the Listeners or
the NetworkPolicy were not fully updated yet. Now we rely solely on the
bpf policy map sync that is done in regenerateBPF() after the wait for
Envoy configuration updates is completed. At that point Envoy is ready to
take traffic for the new listeners. Also after this point the old
redirects would not receive any traffic any more and they can safely be
removed by the finalizer function called via defer at the end of
regenerateBPF().

Finally, remove old redirects through FinalizeList *after* the new
redirects have been acknowledged by Envoy. Since references to the
redirect ports of those redirects have already been removed from the
policy map, there is no need to wait for an ack for redirect
removal. This allows simplification of the related code as FinalizedList
nor RevertStack are no longer needed for redirect removal.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
